reports/M1-REPORT.md

```diff
@@ -0,0 +1,255 @@
+  2. We modify the PureScript compiler to transform the sugared AST into a *typed* variant of PureScript's `CoreFn` AST. The simplest way to do this is to tag expressions with their type in the annotation field, though it may be useful to create our own variant AST for ergonomic reasons.
+  3. We write a trivial code generator that serializes typed `CoreFn` modules and writes them to the output directory. The compiler will also generate suitable `ExternsFile`s, which are needed for incremental compilation and linking. (This is the last step that involves the PS compiler directly)
+    - The set of `CoreFn` modules generated at this stage can serve as a package or library which may be shared
+  4. We implement a second-pass compiler executable which:
+    1. Parses the `Main` source file (which should contain a `main` function that will be compiled to a UPLC script) and transforms it into typed `CoreFn`
+    2. Parses the `CoreFn` modules and `ExternsFile`s for modules imported by the `Main` source file
+    3. Resolves imports
+    4. Performs inlining
+      1. We probably _have_ to inline row-polymorphic functions because we cannot represent their types in a PIR bindings/declaration
+      2. We probably do not need to inline monomorphic or `Type`-polymorphic functions, for which we can generate PIR bindings/declarations
+  5. Applies all type arguments / instantiated all type variables (throwing an error if a variable cannot be instantiated to a concrete type)
+  6. Eliminates `ObjectUpdate` and `Accessor` expressions using fully-instantiated (closed) row types
+  7. Performs any optimizations
+  8. Transforms the typed `CoreFn` AST into a PIR AST
+  9. Compiles PIR to UPLC
+  10. Seralizes the UPLC Plutus Script and writes it to the output directory
+
+## Possible optimizations
+
+### From purescript-backend-optimizer
+
+[purescript-backend-optimizer](https://github.com/aristanetworks/purescript-backend-optimizer) is a tool that consumes the compiler's high-level IR (CoreFn) and applies a more aggressive inlining pipeline (subsuming existing optimizations) that is backend agnostic.
+
+Some of the optimizations are implemented on the JavaScript level, and are therefore not applicable.
+
+Among those that are implemented on CoreFn level, there is only one:
+
+#### Pattern matching optimization
+
+Naive pattern matching implementations often suffer from multiple dispatching on the same variants many times, because the path information from previous matches are not considered in the following ones. [purescript-backend-optimizer implements](https://github.com/aristanetworks/purescript-backend-optimizer/blob/bdef8b58cd8591bef7fd499c892989604e6b918d/docs/optimized-pattern-matching.md) an algorithm that transforms a single pattern matching statement into a tree of nested pattern matchings, that work faster because a single constructor is never dispatched on more than once.
+
```

Comment:
In LB we print out 'optimal' pattern matching for backends that don't have it supported first class, so yeah, we can manage that at the source level